### DISCUSSION TITLE
[Proposal]: Union types

### DISCUSSION THREAD

**Discussion by savaparoski** [dotnet/csharplang#4730] [Language Ideas]
# Union types

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary
In TypeScript we have union types where we can say that variable ``status`` can be only one of the following values ``"On" | "Off" | "Idle"``

```
type status = "On" | "Off" | "Idle"

const whatsTheStatus: status = "On"; // can be only "On" or "Off" or "Idle"
```

## Motivation
[motivation]: #motivation

## Detailed design
[design]: #detailed-design

## Drawbacks
[drawbacks]: #drawbacks

## Alternatives
[alternatives]: #alternatives

## Unresolved questions
[unresolved]: #unresolved-questions

**Comment by gulshan** [comment]
#399

**Comment by jnm2** [comment]
Moving to a discussion. As mentioned in the issue template:

> Hello, and thanks for your interest in contributing to C#! If you haven't been invited by a team member to open an issue, please instead open a discussion marked [draft issue] at https://github.com/dotnet/csharplang/discussions/new and we'll try to give you feedback on how to get to an issue-ready proposal.
> 
> New language feature proposals should fully fill out this template. This should include a complete detailed design, which describes the syntax of the feature, what that syntax means, and how it affects current parts of the spec. Please make sure to point out specific spec sections that need to be updated for this feature.

**Comment by jnm2** [comment]
@savans2 Union types already being discussed for C# are where the possible values are not constants but are rather instances of a given set of types. A union type would look like `string | string[] | StringBuilder` or `IDisposable | IAsyncDisposable`. @gulshan already linked https://github.com/dotnet/csharplang/discussions/399 and there is a proposal draft at https://github.com/dotnet/csharplang/discussions/4628.

The example given uses constant values, though (`"On" | "Off" | "Idle"`), so looks like it maps to the existing C# concept of an enum. You might be interested in the string-based enum discussion then: https://github.com/dotnet/csharplang/issues/2849
